lib/frame.rb

```diff
@@ -0,0 +1,60 @@
+# knows if it's a strike, spare or over
+class Frame
+  TOTAL_PINS = 10
+
+  attr_reader :remaining_pins, :bonus, :rolls
+
+  def initialize
+    @bonus = 0
+    @rolls = []
+    @final = false
+    @remaining_pins = TOTAL_PINS
+  end
+
+  def score
+    rolls.sum + bonus
+  end
+
+  def add_roll(pins)
+    pins = pins.abs
+    raise PinError if pins > remaining_pins
+
+    self.remaining_pins -= pins
+    replace_pins if final? && !over?
```

Comment:
Is this because instance variables should not be used for multiple purposes like this? or to avoid resetting the variable as you mention below?

I'm not sure I understand exactly, but would something like this be better?
```
  def initialize
    @bonus = 0
    @rolls = []
    @final = false
    @remaining_pins = TOTAL_PINS + ghost_frame
    @ghost_frame = TOTAL_PINS if final?
  end